File: Autonomous_Mobile_Robots_Course/Chapter8/Lesson4/Chapter8_Lesson4.cpp
// Chapter8_Lesson4.cpp
// Particle-Filter Localization — Degeneracy + Kidnapped Robot Recovery (2D)
// Build (example):
//   g++ -O2 -std=c++17 -o Chapter8_Lesson4 Chapter8_Lesson4.cpp
//
// Output:
//   Chapter8_Lesson4_results.csv

#include <algorithm>
#include <cmath>
#include <cstdint>
#include <fstream>
#include <iostream>
#include <random>
#include <vector>

struct Pose {
  double x{0}, y{0}, th{0};
};

struct Particle {
  double x{0}, y{0}, th{0}, w{0};
};

static double wrap_angle(double a) {
  const double pi = 3.14159265358979323846;
  a = std::fmod(a + pi, 2.0 * pi);
  if (a < 0) a += 2.0 * pi;
  return a - pi;
}

static Particle motion_step(const Particle& p, double v, double w, double dt,
                            double sigma_v, double sigma_w,
                            std::mt19937_64& gen) {
  std::normal_distribution<double> nv(0.0, sigma_v);
  std::normal_distribution<double> nw(0.0, sigma_w);

  double v_n = v + nv(gen);
  double w_n = w + nw(gen);

  Particle q = p;
  if (std::abs(w_n) < 1e-9) {
    q.x = p.x + v_n * dt * std::cos(p.th);
    q.y = p.y + v_n * dt * std::sin(p.th);
    q.th = p.th;
  } else {
    q.x = p.x + (v_n / w_n) * (std::sin(p.th + w_n * dt) - std::sin(p.th));
    q.y = p.y - (v_n / w_n) * (std::cos(p.th + w_n * dt) - std::cos(p.th));
    q.th = p.th + w_n * dt;
  }
  q.th = wrap_angle(q.th);
  return q;
}

static void measurement_model(const Pose& x,
                              const std::vector<std::pair<double,double>>& landmarks,
                              std::vector<std::pair<double,double>>& out_rb) {
  out_rb.resize(landmarks.size());
  for (size_t i = 0; i < landmarks.size(); ++i) {
    double dx = landmarks[i].first - x.x;
    double dy = landmarks[i].second - x.y;
    double r = std::sqrt(dx*dx + dy*dy);
    double b = std::atan2(dy, dx) - x.th;
    b = wrap_angle(b);
    out_rb[i] = {r, b};
  }
}

static double log_gauss_0(double e, double sigma) {
  const double pi = 3.14159265358979323846;
  return -0.5 * (e/sigma)*(e/sigma) - std::log(std::sqrt(2.0*pi)*sigma);
}

static double log_sensor_likelihood(const Particle& p,
                                    const std::vector<std::pair<double,double>>& z,
                                    const std::vector<std::pair<double,double>>& landmarks,
                                    double sigma_r, double sigma_b) {
  std::vector<std::pair<double,double>> zh;
  measurement_model(Pose{p.x, p.y, p.th}, landmarks, zh);

  double ll = 0.0;
  for (size_t i = 0; i < z.size(); ++i) {
    double dr = z[i].first  - zh[i].first;
    double db = wrap_angle(z[i].second - zh[i].second);
    ll += log_gauss_0(dr, sigma_r);
    ll += log_gauss_0(db, sigma_b);
  }
  return ll;
}

static void normalize(std::vector<Particle>& P) {
  double s = 0.0;
  for (const auto& p : P) s += p.w;
  if (!(s > 0.0) || !std::isfinite(s)) {
    double w = 1.0 / static_cast<double>(P.size());
    for (auto& p : P) p.w = w;
    return;
  }
  for (auto& p : P) p.w /= s;
}

static double Neff(const std::vector<Particle>& P) {
  double s2 = 0.0;
  for (const auto& p : P) s2 += p.w * p.w;
  if (s2 <= 0.0) return 0.0;
  return 1.0 / s2;
}

static std::vector<Particle> systematic_resample(const std::vector<Particle>& P,
                                                 std::mt19937_64& gen) {
  const size_t N = P.size();
  std::vector<double> cdf(N);
  double acc = 0.0;
  for (size_t i = 0; i < N; ++i) {
    acc += P[i].w;
    cdf[i] = acc;
  }
  cdf[N-1] = 1.0;

  std::uniform_real_distribution<double> ur(0.0, 1.0);
  double u0 = ur(gen) / static_cast<double>(N);

  std::vector<Particle> out;
  out.reserve(N);

  size_t i = 0;
  for (size_t m = 0; m < N; ++m) {
    double u = u0 + static_cast<double>(m) / static_cast<double>(N);
    while (u > cdf[i]) ++i;
    Particle q = P[i];
    q.w = 1.0 / static_cast<double>(N);
    out.push_back(q);
  }
  return out;
}

static void roughen(std::vector<Particle>& P, double k,
                    double xmin, double xmax, double ymin, double ymax,
                    std::mt19937_64& gen) {
  const double pi = 3.14159265358979323846;
  const double N = static_cast<double>(P.size());
  const double d = 3.0;
  double sx = k * (xmax - xmin) * std::pow(N, -1.0/d);
  double sy = k * (ymax - ymin) * std::pow(N, -1.0/d);
  double sth = k * (2.0*pi) * std::pow(N, -1.0/d);

  std::normal_distribution<double> nx(0.0, sx);
  std::normal_distribution<double> ny(0.0, sy);
  std::normal_distribution<double> nth(0.0, sth);

  for (auto& p : P) {
    p.x += nx(gen);
    p.y += ny(gen);
    p.th = wrap_angle(p.th + nth(gen));
  }
}

static void inject_random(std::vector<Particle>& P, double frac,
                          double xmin, double xmax, double ymin, double ymax,
                          std::mt19937_64& gen) {
  const size_t N = P.size();
  size_t m = static_cast<size_t>(std::llround(std::max(0.0, std::min(1.0, frac)) * static_cast<double>(N)));
  if (m == 0) return;

  std::uniform_real_distribution<double> ux(xmin, xmax);
  std::uniform_real_distribution<double> uy(ymin, ymax);
  std::uniform_real_distribution<double> uth(-3.14159265358979323846, 3.14159265358979323846);

  for (size_t i = 0; i < m; ++i) {
    P[i].x = ux(gen);
    P[i].y = uy(gen);
    P[i].th = uth(gen);
  }
}

static Pose estimate(const std::vector<Particle>& P) {
  double x=0.0, y=0.0;
  double c=0.0, s=0.0;
  for (const auto& p : P) {
    x += p.w * p.x;
    y += p.w * p.y;
    c += p.w * std::cos(p.th);
    s += p.w * std::sin(p.th);
  }
  return Pose{x, y, std::atan2(s, c)};
}

int main() {
  std::mt19937_64 gen(4);

  double xmin=0.0, xmax=10.0, ymin=0.0, ymax=10.0;
  std::vector<std::pair<double,double>> landmarks{
    {2.0,2.0},{8.0,2.0},{8.0,8.0},{2.0,8.0}
  };

  double dt = 0.1;
  int T = 300;
  int kidnapped_t = 170;

  double sigma_v = 0.05, sigma_w = 0.03;
  double sigma_r = 0.15, sigma_b = 0.07;

  int N = 800;
  double Neff_ratio = 0.5;
  double rough_k = 0.15;

  double eps_min = 0.01, eps_max = 0.30;
  double ll_thresh = -12.0;

  Pose xtrue{1.0, 1.0, 0.0};

  std::uniform_real_distribution<double> ux(xmin, xmax);
  std::uniform_real_distribution<double> uy(ymin, ymax);
  std::uniform_real_distribution<double> uth(-3.14159265358979323846, 3.14159265358979323846);

  std::vector<Particle> P;
  P.reserve(N);
  for (int i = 0; i < N; ++i) {
    P.push_back(Particle{ux(gen), uy(gen), uth(gen), 1.0/static_cast<double>(N)});
  }

  auto control = [](int t) {
    if (t < 70)  return std::pair<double,double>{0.7, 0.0};
    if (t < 90)  return std::pair<double,double>{0.7, 0.9};
    if (t < 160) return std::pair<double,double>{0.7, 0.0};
    if (t < 180) return std::pair<double,double>{0.7, 0.9};
    if (t < 250) return std::pair<double,double>{0.7, 0.0};
    return std::pair<double,double>{0.7, 0.9};
  };

  std::normal_distribution<double> nr(0.0, sigma_r);
  std::normal_distribution<double> nb(0.0, sigma_b);

  std::ofstream f("Chapter8_Lesson4_results.csv");
  f << "t,true_x,true_y,true_th,est_x,est_y,est_th,Neff,eps\n";

  std::vector<std::pair<double,double>> zhat, z;

  for (int t = 0; t < T; ++t) {
    auto [v_cmd, w_cmd] = control(t);

    Particle tmp{xtrue.x, xtrue.y, xtrue.th, 1.0};
    tmp = motion_step(tmp, v_cmd, w_cmd, dt, sigma_v, sigma_w, gen);
    xtrue = Pose{tmp.x, tmp.y, tmp.th};

    if (t == kidnapped_t) {
      xtrue = Pose{ux(gen), uy(gen), uth(gen)};
    }

    measurement_model(xtrue, landmarks, zhat);
    z = zhat;
    for (auto& rb : z) {
      rb.first  += nr(gen);
      rb.second = wrap_angle(rb.second + nb(gen));
    }

    for (auto& p : P) {
      Particle q = motion_step(p, v_cmd, w_cmd, dt, sigma_v, sigma_w, gen);
      p.x = q.x; p.y = q.y; p.th = q.th;
    }

    std::vector<double> lls(N);
    double llmax = -1e300;
    for (int i = 0; i < N; ++i) {
      lls[i] = log_sensor_likelihood(P[i], z, landmarks, sigma_r, sigma_b);
      llmax = std::max(llmax, lls[i]);
    }
    for (int i = 0; i < N; ++i) {
      P[i].w = std::exp(lls[i] - llmax);
    }
    normalize(P);

    double lbar = 0.0;
    for (int i = 0; i < N; ++i) {
      lbar += P[i].w * log_sensor_likelihood(P[i], z, landmarks, sigma_r, sigma_b);
    }
    double eps = (lbar < ll_thresh) ? eps_max : eps_min;

    double neff = Neff(P);
    if (neff < Neff_ratio * static_cast<double>(N)) {
      P = systematic_resample(P, gen);
      roughen(P, rough_k, xmin, xmax, ymin, ymax, gen);
      inject_random(P, eps, xmin, xmax, ymin, ymax, gen);

      llmax = -1e300;
      for (int i = 0; i < N; ++i) {
        lls[i] = log_sensor_likelihood(P[i], z, landmarks, sigma_r, sigma_b);
        llmax = std::max(llmax, lls[i]);
      }
      for (int i = 0; i < N; ++i) {
        P[i].w = std::exp(lls[i] - llmax);
      }
      normalize(P);
    }

    Pose xest = estimate(P);

    f << t << "," << xtrue.x << "," << xtrue.y << "," << xtrue.th << ","
      << xest.x << "," << xest.y << "," << xest.th << ","
      << neff << "," << eps << "\n";
  }

  std::cout << "Saved: Chapter8_Lesson4_results.csv\n";
  return 0;
}

File: Tutorials/Codes/Autonomous_Mobile_Robots_Course/Chapter8/Lesson4/Chapter8_Lesson4.m
% Chapter8_Lesson4.m
% Particle-Filter Localization — Degeneracy + Kidnapped Robot Recovery (2D)
%
% Run:
%   Chapter8_Lesson4
%
% Notes:
% - This script simulates a 2D robot and range-bearing measurements to known landmarks,
%   runs MCL, monitors degeneracy via N_eff, and recovers from a "kidnapped robot"
%   using adaptive random particle injection.
% - At the end, it writes a CSV file "Chapter8_Lesson4_results.csv".
%
% Optional (Simulink):
% - Call buildSimulinkSkeleton() to auto-create a basic Simulink model with a
%   MATLAB Function block hosting the PF update (lightweight scaffold).

function Chapter8_Lesson4()
    rng(4);

    % Bounds and landmarks
    xmin = 0; xmax = 10; ymin = 0; ymax = 10;
    landmarks = [2 2; 8 2; 8 8; 2 8];

    % Simulation parameters
    dt = 0.1;
    T = 300;
    kidnapped_t = 170;

    % Noise
    sigma_v = 0.05; sigma_w = 0.03;
    sigma_r = 0.15; sigma_b = 0.07;

    % PF settings
    N = 800;
    Neff_ratio = 0.5;
    rough_k = 0.15;

    % Kidnapped recovery
    eps_min = 0.01;
    eps_max = 0.30;
    ll_thresh = -12.0;

    % True state
    x_true = [1.0; 1.0; 0.0];

    % Initialize particles globally
    P = zeros(N,3);
    P(:,1) = xmin + (xmax - xmin) * rand(N,1);
    P(:,2) = ymin + (ymax - ymin) * rand(N,1);
    P(:,3) = -pi + 2*pi*rand(N,1);
    w = ones(N,1)/N;

    true_hist = zeros(T,3);
    est_hist  = zeros(T,3);
    Neff_hist = zeros(T,1);
    eps_hist  = zeros(T,1);

    for t = 1:T
        [v_cmd, w_cmd] = controlProfile(t);

        % True motion (unicycle + noise)
        x_true = motionStep(x_true, v_cmd, w_cmd, dt, sigma_v, sigma_w);

        % Kidnapped event
        if t == kidnapped_t
            x_true = [xmin + (xmax - xmin)*rand();
                      ymin + (ymax - ymin)*rand();
                      -pi + 2*pi*rand()];
        end

        % Measurement from true pose
        z = measurementModel(x_true, landmarks);
        z(:,1) = z(:,1) + sigma_r * randn(size(z,1),1);
        z(:,2) = wrapAngle(z(:,2) + sigma_b * randn(size(z,1),1));

        % Predict particles
        for i = 1:N
            P(i,:) = motionStep(P(i,:).', v_cmd, w_cmd, dt, sigma_v, sigma_w).';
        end

        % Update weights in log domain
        ll = zeros(N,1);
        for i = 1:N
            ll(i) = logSensorLikelihood(P(i,:).', z, landmarks, sigma_r, sigma_b);
        end
        ll = ll - max(ll);
        w = exp(ll);
        w = w / sum(w);

        % Weighted mean log-likelihood for mismatch detection
        ll_w = zeros(N,1);
        for i = 1:N
            ll_w(i) = logSensorLikelihood(P(i,:).', z, landmarks, sigma_r, sigma_b);
        end
        lbar = sum(w .* ll_w);

        if lbar < ll_thresh
            eps = eps_max;
        else
            eps = eps_min;
        end

        % Degeneracy
        Neff = 1 / sum(w.^2);

        if Neff < Neff_ratio * N
            [P, w] = systematicResample(P, w);

            % Roughen
            P = roughen(P, rough_k, [xmin xmax ymin ymax]);

            % Inject random particles
            m = round(eps*N);
            if m > 0
                P(1:m,1) = xmin + (xmax - xmin) * rand(m,1);
                P(1:m,2) = ymin + (ymax - ymin) * rand(m,1);
                P(1:m,3) = -pi + 2*pi*rand(m,1);
            end

            % Re-weight after injection
            ll2 = zeros(N,1);
            for i = 1:N
                ll2(i) = logSensorLikelihood(P(i,:).', z, landmarks, sigma_r, sigma_b);
            end
            ll2 = ll2 - max(ll2);
            w = exp(ll2);
            w = w / sum(w);
        end

        x_est = estimatePose(P, w);

        true_hist(t,:) = x_true.';
        est_hist(t,:)  = x_est.';
        Neff_hist(t) = Neff;
        eps_hist(t)  = eps;
    end

    % Save CSV
    fid = fopen('Chapter8_Lesson4_results.csv','w');
    fprintf(fid, 't,true_x,true_y,true_th,est_x,est_y,est_th,Neff,eps\n');
    for t = 1:T
        fprintf(fid, '%d,%.6f,%.6f,%.6f,%.6f,%.6f,%.6f,%.6f,%.6f\n', ...
            t-1, true_hist(t,1), true_hist(t,2), true_hist(t,3), ...
            est_hist(t,1), est_hist(t,2), est_hist(t,3), ...
            Neff_hist(t), eps_hist(t));
    end
    fclose(fid);
    disp('Saved: Chapter8_Lesson4_results.csv');

    % Plot
    figure; plot(true_hist(:,1), true_hist(:,2), 'LineWidth', 1.5); hold on;
    plot(est_hist(:,1), est_hist(:,2), 'LineWidth', 1.5);
    scatter(landmarks(:,1), landmarks(:,2), 60, 'x', 'LineWidth', 2);
    axis equal; grid on; legend('true','estimate','landmarks');
    title(sprintf('MCL trajectory (kidnapped at t=%d)', kidnapped_t));

    figure; plot(Neff_hist, 'LineWidth', 1.5); hold on;
    yline(Neff_ratio * N, '--');
    grid on; title('Effective sample size N_{eff}');

    figure; plot(eps_hist, 'LineWidth', 1.5);
    grid on; title('Injection fraction eps');

    % Uncomment to generate a Simulink scaffold:
    % buildSimulinkSkeleton();
end

function [v, w] = controlProfile(t)
    if t < 70
        v = 0.7; w = 0.0;
    elseif t < 90
        v = 0.7; w = 0.9;
    elseif t < 160
        v = 0.7; w = 0.0;
    elseif t < 180
        v = 0.7; w = 0.9;
    elseif t < 250
        v = 0.7; w = 0.0;
    else
        v = 0.7; w = 0.9;
    end
end

function x2 = motionStep(x, v, w, dt, sigma_v, sigma_w)
    v_n = v + sigma_v * randn();
    w_n = w + sigma_w * randn();
    if abs(w_n) < 1e-9
        x2 = [x(1) + v_n*dt*cos(x(3));
              x(2) + v_n*dt*sin(x(3));
              x(3)];
    else
        x2 = [x(1) + (v_n/w_n)*(sin(x(3)+w_n*dt) - sin(x(3)));
              x(2) - (v_n/w_n)*(cos(x(3)+w_n*dt) - cos(x(3)));
              x(3) + w_n*dt];
    end
    x2(3) = wrapAngle(x2(3));
end

function z = measurementModel(x, landmarks)
    dx = landmarks(:,1) - x(1);
    dy = landmarks(:,2) - x(2);
    r = sqrt(dx.^2 + dy.^2);
    b = atan2(dy, dx) - x(3);
    b = wrapAngle(b);
    z = [r, b];
end

function ll = logSensorLikelihood(xp, z, landmarks, sigma_r, sigma_b)
    zh = measurementModel(xp, landmarks);
    dr = z(:,1) - zh(:,1);
    db = wrapAngle(z(:,2) - zh(:,2));
    cr = -log(sqrt(2*pi)*sigma_r);
    cb = -log(sqrt(2*pi)*sigma_b);
    ll = sum(-0.5*(dr./sigma_r).^2 + cr) + sum(-0.5*(db./sigma_b).^2 + cb);
end

function a = wrapAngle(a)
    a = mod(a + pi, 2*pi) - pi;
end

function [P2, w2] = systematicResample(P, w)
    N = size(P,1);
    cdf = cumsum(w);
    cdf(end) = 1.0;
    u0 = rand()/N;
    idx = zeros(N,1);
    i = 1;
    for m = 1:N
        u = u0 + (m-1)/N;
        while u > cdf(i)
            i = i + 1;
        end
        idx(m) = i;
    end
    P2 = P(idx,:);
    w2 = ones(N,1)/N;
end

function P2 = roughen(P, k, bounds)
    xmin = bounds(1); xmax = bounds(2);
    ymin = bounds(3); ymax = bounds(4);
    N = size(P,1);
    d = 3;
    sx  = k*(xmax-xmin)*N^(-1/d);
    sy  = k*(ymax-ymin)*N^(-1/d);
    sth = k*(2*pi)*N^(-1/d);
    P2 = P;
    P2(:,1) = P(:,1) + sx*randn(N,1);
    P2(:,2) = P(:,2) + sy*randn(N,1);
    P2(:,3) = wrapAngle(P(:,3) + sth*randn(N,1));
end

function xhat = estimatePose(P, w)
    x = sum(w .* P(:,1));
    y = sum(w .* P(:,2));
    c = sum(w .* cos(P(:,3)));
    s = sum(w .* sin(P(:,3)));
    th = atan2(s, c);
    xhat = [x; y; th];
end

function buildSimulinkSkeleton()
    model = 'Chapter8_Lesson4_MCL_Simulink';
    if bdIsLoaded(model)
        close_system(model, 0);
    end
    new_system(model); open_system(model);

    add_block('simulink/Sources/Constant', [model '/u_vw'], 'Value', '[0.7; 0.0]');
    add_block('simulink/Sources/Constant', [model '/z_meas'], 'Value', 'zeros(4,2)');

    add_block('simulink/User-Defined Functions/MATLAB Function', [model '/MCL_Update']);
    set_param([model '/MCL_Update'], 'Script', sprintf([ ...
        'function xhat = f(u, z)\n' ...
        '%% u = [v; w], z = measurements\n' ...
        '%% This block is a scaffold. Paste your PF predict/update/resample here.\n' ...
        'xhat = zeros(3,1);\n' ...
        'end\n' ...
    ]));

    add_block('simulink/Sinks/Display', [model '/xhat_display']);

    add_line(model, 'u_vw/1', 'MCL_Update/1');
    add_line(model, 'z_meas/1', 'MCL_Update/2');
    add_line(model, 'MCL_Update/1', 'xhat_display/1');

    set_param(model, 'StopTime', '10');
    save_system(model);
    disp(['Created Simulink scaffold: ' model]);
end
